File: src/TelepathyLabs.ShowReels.Domain/VideoClip.cs
namespace TelepathyLabs.ShowReels.Domain
{
    public class VideoClip
    {
        public VideoClip(
            string name,
            string description,
            VideoStandard videoStandard,
            VideoDefinition videoDefinition,
            TimeCode startTimeCode,
            TimeCode endTimeCode)
        {
            if (string.IsNullOrWhiteSpace(name))
            {
                throw new ShowReelsException("Name cannot be null or empty.");
            }

            if (string.IsNullOrWhiteSpace(description))
            {
                throw new ShowReelsException("Description cannot be null or empty.");
            }

            if (startTimeCode == null)
            {
                throw new ShowReelsException("Start timecode cannot be null.");
            }

            if (endTimeCode == null)
            {
                throw new ShowReelsException("End timecode cannot be null.");
            }

            if (startTimeCode.FramesPerSecond != endTimeCode.FramesPerSecond)
            {
                throw new ShowReelsException("Frame rates for start time and end time should match.");
            }

            if (startTimeCode > endTimeCode)
            {
                throw new ShowReelsException("Start time should be smaller than end time.");
            }

            Name = name;
            Description = description;
            VideoDefinition = videoDefinition;
            VideoStandard = videoStandard;
            StartTimeCode = startTimeCode;
            EndTimeCode = endTimeCode;
        }

        public int Id { get; set; }
        public string Name { get; private set; }
        public string Description { get; private set; }
        public VideoStandard VideoStandard { get; private set; }
        public VideoDefinition VideoDefinition { get; private set; }
        public TimeCode StartTimeCode { get; private set; }
        public TimeCode EndTimeCode { get; private set; }
    }
}

File: src/TelepathyLabs.ShowReels.Web/src/app/services/models/video-clip.ts
import { TimeCode } from "./time-code";

export class VideoClip {

    constructor(
        name: string,
        description: string,
        videoDefinition: number,
        videoStandard: number,
        startTimeCode: TimeCode,
        endTimeCode: TimeCode
    ) {
        if (!name)
        {
            throw new Error("Name cannot be null or empty.");
        }

        if (!description)
        {
            throw new Error("Description cannot be null or empty.");
        }

        if (startTimeCode === null)
        {
            throw new Error("Start timecode cannot be null.");
        }

        if (endTimeCode === null)
        {
            throw new Error("End timecode cannot be null.");
        }

        if (startTimeCode.FramesPerSecond !== endTimeCode.FramesPerSecond)
        {
            throw new Error("Frame rates for start time and end time should match.");
        }

        if (startTimeCode.IsLargerThan(endTimeCode) || startTimeCode.IsEqualTo(endTimeCode))
        {
            throw new Error("Start time should be smaller than end time.");
        }

        this.name = name;
        this.description = description;
        this.videoDefinition = videoDefinition;
        this.videoStandard = videoStandard;
        this.startTimeCode = startTimeCode;
        this.endTimeCode = endTimeCode;
    }

    private name: string;
    public get Name(){
        return this.name;
    }

    private description: string;
    public get Description(){
        return this.description;
    }

    private videoDefinition: number;
    public get VideoDefinition(){
        return this.videoDefinition;
    }

    private videoStandard: number;
    public get VideoStandard(){
        return this.videoStandard;
    }

    private startTimeCode: TimeCode;
    public get StartTimeCode(){
        return this.startTimeCode;
    }

    private endTimeCode: TimeCode;
    public get EndTimeCode(){
        return this.endTimeCode;
    }
}